cpp/src/arrow/compute/expression.h

```diff
@@ -0,0 +1,261 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "arrow/util/macros.h"
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+
+class Status;
+
+namespace compute {
+
+class LogicalType;
+class ExprVisitor;
+class Operation;
+
+/// \brief Base class for all analytic expressions. Expressions may represent
+/// data values (scalars, arrays, tables)
+class ARROW_EXPORT Expr {
+ public:
+  /// \brief Instantiate expression from an abstract operation
+  /// \param[in] op the operation that generates the expression
+  explicit Expr(std::shared_ptr<const Operation> op);
+
+  virtual ~Expr() = default;
+
+  /// \brief A unique string identifier for the kind of expression
+  virtual std::string kind() const = 0;
+
+  /// \brief Accept expression visitor
+  /// TODO(wesm)
+  // virtual Status Accept(ExprVisitor* visitor) const = 0;
+
+  /// \brief The underlying operation
+  std::shared_ptr<const Operation> op() const { return op_; }
+
+ protected:
+  std::shared_ptr<const Operation> op_;
+};
+
+/// \brief Base class for a data-generated expression with a fixed and known
+/// type. This includes arrays and scalars
+class ARROW_EXPORT ValueExpr : public Expr {
+ public:
+  /// The value cardinality: one or many. These correspond to the arrow::Scalar
+  /// and arrow::Array types
+  enum Rank { SCALAR, ARRAY };
+  /// \brief The name of the expression, if any. The default is unnamed
+  // virtual const ExprName& name() const;
+
+  std::shared_ptr<LogicalType> type() const;
+
+  /// \brief The value cardinality (scalar or array) of the expression
+  Rank rank() const { return rank_; }
+
+ protected:
+  ValueExpr(std::shared_ptr<const Operation> op, std::shared_ptr<LogicalType> type,
+            Rank rank);
+
+  /// \brief The semantic data type of the expression
+  std::shared_ptr<LogicalType> type_;
+
+  Rank rank_;
```

Comment:
This is known at compile time via ArrayExpr and ScalarExpr, does it need to be a runtime member?

Reply:
I believe in this case `Rank` represents another `kind`/`type_id`/... -like member; it's the discriminant which allows you to visit an ArrayExpr or ScalarExpr

Reply:
Let me see if I can make `rank()` virtual

Reply:
done